### PR TITLE
Move health check out of transaction scope

### LINKED ISSUE #280: Transaction handling when doing the health check
When doing upgrade, the appDeployerData is persisted within the same transaction that the health check also uses to detect the health of the apps being deployed.

This makes the status check fail (with no appDeployerData for this release) until the health check is complete.

### PR DESCRIPTION
**Pull request by markpollack**
* Break UpdateStrategy into additional steps

Fixes #280

### CODE REVIEW

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+			logger.error("New release " + replacingRelease.getName() + " was not detected as healthy after " +
+					this.healthCheckProperties.getTimeoutInMillis() + "milliseconds.  " +
+					"Keeping existing release, and Deleting apps of replacing release");
+			this.releaseManager.delete(replacingRelease);
+			Status status = new Status();
+			status.setStatusCode(StatusCode.FAILED);
+			replacingRelease.getInfo().setStatus(status);
+			replacingRelease.getInfo().setStatus(status);
+			replacingRelease.getInfo().setDescription("Did not detect apps in repalcing release as healthy after " +
+					this.healthCheckProperties.getSleepInMillis() + " ms.");
```
Comment (ilayaperumalg) [inline spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java]: I think this needs to be `getTimeoutInMillis() `

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+		try {
+			logger.error("New release " + replacingRelease.getName() + " was not detected as healthy after " +
+					this.healthCheckProperties.getTimeoutInMillis() + "milliseconds.  " +
+					"Keeping existing release, and Deleting apps of replacing release");
+			this.releaseManager.delete(replacingRelease);
+			Status status = new Status();
+			status.setStatusCode(StatusCode.FAILED);
+			replacingRelease.getInfo().setStatus(status);
+			replacingRelease.getInfo().setStatus(status);
+			replacingRelease.getInfo().setDescription("Did not detect apps in repalcing release as healthy after " +
```
Comment (ilayaperumalg) [inline spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java]: typo: replacing

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+	private ReleaseManager releaseManager;
+
+	public HandleHealthCheckStep(ReleaseRepository releaseRepository,
+			DeployerRepository deployerRepository,
+			AppDeployerDataRepository appDeployerDataRepository,
+			DeleteStep deleteStep,
+			HealthCheckProperties healthCheckProperties) {
+		this.releaseRepository = releaseRepository;
+		this.deployerRepository = deployerRepository;
```
Comment (ilayaperumalg) [inline spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java]: deployerRepository is not used in this class and hence can be removed.

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+/**
+ * Responsible for taking action based on the health of the latest deployed release. If
+ * healthy, then delete applications from the previous release. Otherwise delete the
+ * latest deployed release. Delegates to {@link DeleteStep}.
+ *
+ * @author Mark Pollack
+ * @author Ilayaperumal Gopinathan
+ */
+public class HandleHealthCheckStep {
```
Comment (ilayaperumalg) [inline spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java]: Maybe `HealthCheckStepHandler` a better name?

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/SimpleRedBlackUpgradeStrategy.java [hunk 5]
```diff
@@ -53,66 +48,42 @@
...
-	private final HealthCheckAndDeleteStep healthCheckAndDeleteStep;
+	private final HealthCheckStep healthCheckStep;
+
+	private final HandleHealthCheckStep handleHealthCheckStep;
+
+	private final ReleaseRepository releaseRepository;
+
+	private final DeployAppStep deployAppStep;
 
 	public SimpleRedBlackUpgradeStrategy(DeployerRepository deployerRepository,
```
Comment (ilayaperumalg) [inline spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/SimpleRedBlackUpgradeStrategy.java]: After simplification, there are dependencies such as `deployerRepository`, `appDeployerDataRepository`, `appDeploymentRequestFactory` and `releaseRepository` are not needed in this class.

**Comment by jvalkeal** [comment]
I think this looks good for now.

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+		logger.info("Release {}-v{} has been DEPLOYED", replacingRelease.getName(),
+				replacingRelease.getVersion());
+		logger.info("Apps in release {}-v{} are healthy.", replacingRelease.getName(),
+				replacingRelease.getVersion());
+	}
+
+	private void deleteReplacingRelease(Release replacingRelease) {
+		try {
+			logger.error("New release " + replacingRelease.getName() + " was not detected as healthy after " +
+					this.healthCheckProperties.getTimeoutInMillis() + "milliseconds.  " +
```
Comment (ilayaperumalg) [inline spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java]: space before milliseconds.

**Comment by ilayaperumalg** [comment]
Apart from the code comments, the changes LGTM as well.

Below are some scenarios tested and they all look good!

```
skipper:>install --package-name log --package-version 1.0.0 --release-name mylog
Released mylog
skipper:>upgrade --package-name log --package-version 1.1.0 --release-name mylog
mylog has been upgraded.
Last Deployed: Thu Oct 26 22:57:52 IST 2017

skipper:>status --release-name mylog
╔═══════════════╤══════════════════════════════════════╗
║Last Deployed  │Thu Oct 26 22:57:52 IST 2017          ║
║Status         │UNKNOWN                               ║
║Platform Status│The applications are being deployed.  ║
║               │mylog.log-v2[mylog.log-v2-0=deploying]║
╚═══════════════╧══════════════════════════════════════╝

skipper:>status --release-name mylog
╔═══════════════╤═════════════════════════════════════════════════╗
║Last Deployed  │Thu Oct 26 22:57:52 IST 2017                     ║
║Status         │DEPLOYED                                         ║
║Platform Status│All applications have been successfully deployed.║
║               │mylog.log-v2[mylog.log-v2-0=deployed]            ║
╚═══════════════╧═════════════════════════════════════════════════╝
skipper:>upgrade --package-name log --package-version 1.1.0 --properties spec.deploymentProperties.spring.cloud.deployer.memory=512m --release-name mylog
mylog has been upgraded.
Last Deployed: Thu Oct 26 22:59:25 IST 2017

skipper:>manifest get --release-name mylog

---
# Source: log.yml
apiVersion: skipper/v1
kind: SpringBootApp
metadata:
  name: log
  type: sink
spec:
  resource: maven://org.springframework.cloud.stream.app:log-sink-rabbit:1.3.0.M1
  resourceMetadata: maven://org.springframework.cloud.stream.app:log-sink-rabbit:jar:metadata:1.3.0.M1
  applicationProperties:
  deploymentProperties:
    spring.cloud.deployer.memory: 512m
    spring.cloud.deployer.count: 1

skipper:>status --release-name mylog
╔═══════════════╤══════════════════════════════════════╗
║Last Deployed  │Thu Oct 26 22:59:25 IST 2017          ║
║Status         │UNKNOWN                               ║
║Platform Status│The applications are being deployed.  ║
║               │mylog.log-v3[mylog.log-v3-0=deploying]║
╚═══════════════╧══════════════════════════════════════╝

skipper:>status --release-name mylog
╔═══════════════╤═════════════════════════════════════════════════╗
║Last Deployed  │Thu Oct 26 22:59:25 IST 2017                     ║
║Status         │DEPLOYED                                         ║
║Platform Status│All applications have been successfully deployed.║
║               │mylog.log-v3[mylog.log-v3-0=deployed]            ║
╚═══════════════╧═════════════════════════════════════════════════╝

skipper:>history --release-name mylog
╔═══════╤════════════════════════════╤════════╤════════════╤═══════════════╤════════════════╗
║Version│        Last updated        │ Status │Package Name│Package Version│  Description   ║
╠═══════╪════════════════════════════╪════════╪════════════╪═══════════════╪════════════════╣
║3      │Thu Oct 26 22:59:25 IST 2017│DEPLOYED│log         │1.1.0          │Upgrade complete║
║2      │Thu Oct 26 22:57:52 IST 2017│DELETED │log         │1.1.0          │Delete complete ║
║1      │Thu Oct 26 22:57:17 IST 2017│DELETED │log         │1.0.0          │Delete complete ║
╚═══════╧════════════════════════════╧════════╧════════════╧═══════════════╧════════════════╝

skipper:>manifest get --release-name mylog --release-version 2

---
# Source: log.yml
apiVersion: skipper/v1
kind: SpringBootApp
metadata:
  name: log
  type: sink
spec:
  resource: maven://org.springframework.cloud.stream.app:log-sink-rabbit:1.3.0.M1
  resourceMetadata: maven://org.springframework.cloud.stream.app:log-sink-rabbit:jar:metadata:1.3.0.M1
  applicationProperties:
  deploymentProperties:

skipper:>rollback --release-name mylog
mylog has been rolled back.
Last Deployed: Thu Oct 26 23:01:25 IST 2017

skipper:>status --release-name mylog
╔═══════════════╤══════════════════════════════════════╗
║Last Deployed  │Thu Oct 26 23:01:25 IST 2017          ║
║Status         │UNKNOWN                               ║
║Platform Status│The applications are being deployed.  ║
║               │mylog.log-v4[mylog.log-v4-0=deploying]║
╚═══════════════╧══════════════════════════════════════╝

skipper:>manifest get --release-name mylog

---
# Source: log.yml
apiVersion: skipper/v1
kind: SpringBootApp
metadata:
  name: log
  type: sink
spec:
  resource: maven://org.springframework.cloud.stream.app:log-sink-rabbit:1.3.0.M1
  resourceMetadata: maven://org.springframework.cloud.stream.app:log-sink-rabbit:jar:metadata:1.3.0.M1
  applicationProperties:
  deploymentProperties:

skipper:>status --release-name mylog
╔═══════════════╤═════════════════════════════════════════════════╗
║Last Deployed  │Thu Oct 26 23:01:25 IST 2017                     ║
║Status         │DEPLOYED                                         ║
║Platform Status│All applications have been successfully deployed.║
║               │mylog.log-v4[mylog.log-v4-0=deployed]            ║
╚═══════════════╧═════════════════════════════════════════════════╝

skipper:>history --release-name mylog
╔═══════╤════════════════════════════╤════════╤════════════╤═══════════════╤════════════════╗
║Version│        Last updated        │ Status │Package Name│Package Version│  Description   ║
╠═══════╪════════════════════════════╪════════╪════════════╪═══════════════╪════════════════╣
║4      │Thu Oct 26 23:01:25 IST 2017│DEPLOYED│log         │1.1.0          │Upgrade complete║
║3      │Thu Oct 26 22:59:25 IST 2017│DELETED │log         │1.1.0          │Delete complete ║
║2      │Thu Oct 26 22:57:52 IST 2017│DELETED │log         │1.1.0          │Delete complete ║
║1      │Thu Oct 26 22:57:17 IST 2017│DELETED │log         │1.0.0          │Delete complete ║
╚═══════╧════════════════════════════╧════════╧════════════╧═══════════════╧════════
skipper:>manifest get --release-name mylog

---
# Source: log.yml
apiVersion: skipper/v1
kind: SpringBootApp
metadata:
  name: log
  type: sink
spec:
  resource: maven://org.springframework.cloud.stream.app:log-sink-rabbit:1.3.0.M1
  resourceMetadata: maven://org.springframework.cloud.stream.app:log-sink-rabbit:jar:metadata:1.3.0.M1
  applicationProperties:
  deploymentProperties:

skipper:>manifest get --release-name mylog --release-version 3

---
# Source: log.yml
apiVersion: skipper/v1
kind: SpringBootApp
metadata:
  name: log
  type: sink
spec:
  resource: maven://org.springframework.cloud.stream.app:log-sink-rabbit:1.3.0.M1
  resourceMetadata: maven://org.springframework.cloud.stream.app:log-sink-rabbit:jar:metadata:1.3.0.M1
  applicationProperties:
  deploymentProperties:
    spring.cloud.deployer.memory: 512m
    spring.cloud.deployer.count: 1

skipper:>rollback --release-name mylog
mylog has been rolled back.
Last Deployed: Thu Oct 26 23:02:52 IST 2017

skipper:>manifest get --release-name mylog

---
# Source: log.yml
apiVersion: skipper/v1
kind: SpringBootApp
metadata:
  name: log
  type: sink
spec:
  resource: maven://org.springframework.cloud.stream.app:log-sink-rabbit:1.3.0.M1
  resourceMetadata: maven://org.springframework.cloud.stream.app:log-sink-rabbit:jar:metadata:1.3.0.M1
  applicationProperties:
  deploymentProperties:
    spring.cloud.deployer.memory: 512m
    spring.cloud.deployer.count: 1

skipper:>status --release-name mylog
╔═══════════════╤══════════════════════════════════════╗
║Last Deployed  │Thu Oct 26 23:02:52 IST 2017          ║
║Status         │UNKNOWN                               ║
║Platform Status│The applications are being deployed.  ║
║               │mylog.log-v5[mylog.log-v5-0=deploying]║
╚═══════════════╧══════════════════════════════════════╝

skipper:>status --release-name mylog
╔═══════════════╤═════════════════════════════════════════════════╗
║Last Deployed  │Thu Oct 26 23:02:52 IST 2017                     ║
║Status         │DEPLOYED                                         ║
║Platform Status│All applications have been successfully deployed.║
║               │mylog.log-v5[mylog.log-v5-0=deployed]            ║
╚═══════════════╧═════════════════════════════════════════════════╝

skipper:>manifest get --release-name mylog

---
# Source: log.yml
apiVersion: skipper/v1
kind: SpringBootApp
metadata:
  name: log
  type: sink
spec:
  resource: maven://org.springframework.cloud.stream.app:log-sink-rabbit:1.3.0.M1
  resourceMetadata: maven://org.springframework.cloud.stream.app:log-sink-rabbit:jar:metadata:1.3.0.M1
  applicationProperties:
  deploymentProperties:
    spring.cloud.deployer.memory: 512m
    spring.cloud.deployer.count: 1

skipper:>

```

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java (hunk 1 above)
Comment (markpollack) [reply spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java]: yes!

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java (hunk 3 above)
Comment (markpollack) [reply spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java]: great, splitting up the steps made the # of dep classes go down, a good thing.

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java (hunk 4 above)
Comment (markpollack) [reply spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HandleHealthCheckStep.java]: I like keeping the 'step' suffix....

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/SimpleRedBlackUpgradeStrategy.java (hunk 5 above)
Comment (markpollack) [reply spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/SimpleRedBlackUpgradeStrategy.java]: can you remove them on merge?

**Comment by ilayaperumalg** [comment]
and some package upgrade failure scenarios:

```
skipper:>install --package-name log --package-version 1.0.0 --release-name mylog
Released mylog
skipper:>upgrade --release-name mylog --package-name log --package-version 1.1.0
mylog has been upgraded.
Last Deployed: Thu Oct 26 22:29:17 IST 2017

skipper:>status --release-name mylog
╔═══════════════╤══════════════════════════════════════╗
║Last Deployed  │Thu Oct 26 22:29:17 IST 2017          ║
║Status         │UNKNOWN                               ║
║Platform Status│The applications are being deployed.  ║
║               │mylog.log-v2[mylog.log-v2-0=deploying]║
╚═══════════════╧══════════════════════════════════════╝

skipper:>status --release-name mylog
╔═══════════════╤══════════════════════════════════════╗
║Last Deployed  │Thu Oct 26 22:29:17 IST 2017          ║
║Status         │UNKNOWN                               ║
║Platform Status│The applications are being deployed.  ║
║               │mylog.log-v2[mylog.log-v2-0=deploying]║
╚═══════════════╧══════════════════════════════════════╝

skipper:>status --release-name mylog
╔═══════════════╤══════════════════════════════════════╗
║Last Deployed  │Thu Oct 26 22:29:17 IST 2017          ║
║Status         │UNKNOWN                               ║
║Platform Status│The applications are being deployed.  ║
║               │mylog.log-v2[mylog.log-v2-0=deploying]║
╚═══════════════╧══════════════════════════════════════╝

skipper:>history --release-name mylog
╔═══════╤════════════════════════════╤════════╤════════════╤═══════════════╤════════════════════════╗
║Version│        Last updated        │ Status │Package Name│Package Version│      Description       ║
╠═══════╪════════════════════════════╪════════╪════════════╪═══════════════╪════════════════════════╣
║2      │Thu Oct 26 22:29:17 IST 2017│UNKNOWN │log         │1.1.0          │Upgrade install underway║
║1      │Thu Oct 26 22:28:58 IST 2017│DEPLOYED│log         │1.0.0          │Install complete        ║
╚═══════╧════════════════════════════╧════════╧════════════╧═══════════════╧════════════════════════╝

skipper:>history --release-name mylog
╔═══════╤════════════════════════════╤════════╤════════════╤═══════════════╤════════════════╗
║Version│        Last updated        │ Status │Package Name│Package Version│  Description   ║
╠═══════╪════════════════════════════╪════════╪════════════╪═══════════════╪════════════════╣
║2      │Thu Oct 26 22:29:17 IST 2017│DEPLOYED│log         │1.1.0          │Upgrade complete║
║1      │Thu Oct 26 22:28:58 IST 2017│DELETED │log         │1.0.0          │Delete complete ║
╚═══════╧════════════════════════════╧════════╧════════════╧═══════════════╧════════════════╝

skipper:>rollback --release-name mylog
mylog has been rolled back.
Last Deployed: Thu Oct 26 22:30:44 IST 2017

skipper:>history --release-name mylog
╔═══════╤════════════════════════════╤════════╤════════════╤═══════════════╤════════════════════════╗
║Version│        Last updated        │ Status │Package Name│Package Version│      Description       ║
╠═══════╪════════════════════════════╪════════╪════════════╪═══════════════╪════════════════════════╣
║3      │Thu Oct 26 22:30:44 IST 2017│UNKNOWN │log         │1.0.0          │Initial install underway║
║2      │Thu Oct 26 22:29:17 IST 2017│DEPLOYED│log         │1.1.0          │Upgrade complete        ║
║1      │Thu Oct 26 22:28:58 IST 2017│DELETED │log         │1.0.0          │Delete complete         ║
╚═══════╧════════════════════════════╧════════╧════════════╧═══════════════╧════════════════════════╝

skipper:>rollback --release-name mylog
mylog has been rolled back.
Last Deployed: Thu Oct 26 22:31:39 IST 2017

skipper:>history --release-name mylog
╔═══════╤════════════════════════════╤════════╤════════════╤═══════════════╤════════════════════════╗
║Version│        Last updated        │ Status │Package Name│Package Version│      Description       ║
╠═══════╪════════════════════════════╪════════╪════════════╪═══════════════╪════════════════════════╣
║4      │Thu Oct 26 22:31:39 IST 2017│UNKNOWN │log         │1.1.0          │Initial install underway║
║3      │Thu Oct 26 22:30:44 IST 2017│DEPLOYED│log         │1.0.0          │Upgrade complete        ║
║2      │Thu Oct 26 22:29:17 IST 2017│DELETED │log         │1.1.0          │Delete complete         ║
║1      │Thu Oct 26 22:28:58 IST 2017│DELETED │log         │1.0.0          │Delete complete         ║
╚═══════╧════════════════════════════╧════════╧════════════╧═══════════════╧════════════════════════╝

skipper:>install --package-name log --package-version 1.0.0 --release-name mylog
Released mylog
skipper:>upgrade --release-name mylog --package-name log --package-version 1.1.0
mylog has been upgraded.
Last Deployed: Thu Oct 26 22:39:20 IST 2017

skipper:>status --release-name mylog
╔═══════════════╤════════════════════════════════════════════════════════════════════════╗
║Last Deployed  │Thu Oct 26 22:39:20 IST 2017                                            ║
║Status         │FAILED                                                                  ║
║Platform Status│The applications are known to the system, but is not currently deployed.║
║               │mylog.log-v2]                                                           ║
╚═══════════════╧════════════════════════════════════════════════════════════════════════╝

skipper:>history --release-name mylog
╔═══════╤════════════════════════════╤════════╤════════════╤═══════════════╤══════════════════════════════════════════════════════════════════╗
║Version│        Last updated        │ Status │Package Name│Package Version│                           Description                            ║
╠═══════╪════════════════════════════╪════════╪════════════╪═══════════════╪══════════════════════════════════════════════════════════════════╣
║2      │Thu Oct 26 22:39:20 IST 2017│FAILED  │log         │1.1.0          │Did not detect apps in repalcing release as healthy after 5000 ms.║
║1      │Thu Oct 26 22:39:10 IST 2017│DEPLOYED│log         │1.0.0          │Install complete                                                  ║
╚═══════╧════════════════════════════╧════════╧════════════╧═══════════════╧══════════════════════════════════════════════════════════════════╝

```

**Comment by ilayaperumalg** [comment]
LGTM, merging.

**Comment by ilayaperumalg** [comment]
Squashed, rebased and merged as  a86afc8a39854d8427e35753edcd9e372774c605 along with addressing the review comments 0fe660571f4fa91e4f37168efe5369511cfb67f9

**Pull request closed without merging**